channeldb/invoices.go

```diff
@@ -135,90 +137,129 @@ const (
+// avaialbe value for the AddIndex field.
+//
+// NOTE: This function will be executed in its own transaction and will
+// use the default timeout.
+func (d *DB) AddInvoice(invoice *invpkg.Invoice) error {
+	tx, err := d.BeginReadWriteTx()
+	if err != nil {
+		return err
+	}
 
-	if err := invpkg.ValidateInvoice(newInvoice, paymentHash); err != nil {
-		return 0, err
+	// Make sure the transaction rolls back in the event of a panic.
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	ctxt, cancel := context.WithTimeout(
+		context.Background(), database.DefaultStoreTimeout,
+	)
+	defer cancel()
+
+	err = d.AddInvoiceWithTx(ctxt, tx, invoice)
+	if err != nil {
+		return err
 	}
 
-	var invoiceAddIndex uint64
-	err := kvdb.Update(d, func(tx kvdb.RwTx) error {
-		invoices, err := tx.CreateTopLevelBucket(invoiceBucket)
-		if err != nil {
-			return err
-		}
+	return tx.Commit()
+}
 
-		invoiceIndex, err := invoices.CreateBucketIfNotExists(
-			invoiceIndexBucket,
-		)
-		if err != nil {
-			return err
-		}
-		addIndex, err := invoices.CreateBucketIfNotExists(
-			addIndexBucket,
-		)
-		if err != nil {
-			return err
-		}
+// AddInvoiceWithTx inserts the targeted invoice into the database.
+//
+// NOTE: If the invoice is added to the database this will set the next
+// available value for the AddIndex field.
+//
+// NOTE: This method is part of the InvoiceDB interface.
+func (d *DB) AddInvoiceWithTx(ctx context.Context, dbTx database.DBTx,
+	invoice *invpkg.Invoice) error {
 
-		// Ensure that an invoice an identical payment hash doesn't
-		// already exist within the index.
-		if invoiceIndex.Get(paymentHash[:]) != nil {
-			return invpkg.ErrDuplicateInvoice
-		}
+	tx, ok := dbTx.(kvdb.RwTx)
+	if !ok {
+		return errors.New("invalid transaction type")
+	}
 
-		// Check that we aren't inserting an invoice with a duplicate
-		// payment address. The all-zeros payment address is
-		// special-cased to support legacy keysend invoices which don't
-		// assign one. This is safe since later we also will avoid
-		// indexing them and avoid collisions.
-		payAddrIndex := tx.ReadWriteBucket(payAddrIndexBucket)
-		if newInvoice.Terms.PaymentAddr != invpkg.BlankPayAddr {
-			paymentAddr := newInvoice.Terms.PaymentAddr[:]
-			if payAddrIndex.Get(paymentAddr) != nil {
-				return invpkg.ErrDuplicatePayAddr
-			}
-		}
+	var hash lntypes.Hash
+	switch {
+	case invoice.Terms.PaymentPreimage != nil:
+		hash = invoice.Terms.PaymentPreimage.Hash()
 
-		// If the current running payment ID counter hasn't yet been
-		// created, then create it now.
-		var invoiceNum uint32
-		invoiceCounter := invoiceIndex.Get(numInvoicesKey)
-		if invoiceCounter == nil {
-			var scratch [4]byte
-			byteOrder.PutUint32(scratch[:], invoiceNum)
-			err := invoiceIndex.Put(numInvoicesKey, scratch[:])
-			if err != nil {
-				return err
-			}
-		} else {
-			invoiceNum = byteOrder.Uint32(invoiceCounter)
+	case invoice.Terms.PaymentHash != nil:
```

Comment:
nit: I think this case should come first so we don't recompute the hash from the preimage all the time if the preimage is set.